libs/langchain/langchain/retrievers/self_query/supabase.py

```diff
@@ -0,0 +1,96 @@
+from typing import Any, Dict, Tuple
+
+from langchain.chains.query_constructor.ir import (
+    Comparator,
+    Comparison,
+    Operation,
+    Operator,
+    StructuredQuery,
+    Visitor,
+)
+
+
+class SupabaseVectorTranslator(Visitor):
+    """Translate Langchain filters to Supabase PostgREST filters."""
+
+    allowed_operators = [Operator.AND, Operator.OR]
+    """Subset of allowed logical operators."""
+    allowed_comparators = [
+        Comparator.EQ,
+        Comparator.NE,
+        Comparator.GT,
+        Comparator.GTE,
+        Comparator.LT,
+        Comparator.LTE,
+        Comparator.CONTAIN,
+        Comparator.LIKE,
+    ]
+    """Subset of allowed logical comparators."""
+
+    metadata_column = "metadata"
+
+    def _format_value(self, value: Any, comparator: Comparator) -> str:
+        if comparator == Comparator.CONTAIN:
+            return f"%{value}%"
+
+        return value
+
+    def _map_comparator(self, comparator: Comparator) -> str:
+        """
+        Maps Langchain comparator to PostgREST comparator:
+
+        https://postgrest.org/en/stable/references/api/tables_views.html#operators
+        """
+        return {
+            Comparator.EQ: "eq",
+            Comparator.NE: "neq",
+            Comparator.GT: "gt",
+            Comparator.GTE: "gte",
+            Comparator.LT: "lt",
+            Comparator.LTE: "lte",
+            Comparator.CONTAIN: "like",
+            Comparator.LIKE: "like",
+        }.get(comparator, comparator)
+
+    def _get_json_operator(self, value: Any) -> str:
+        if isinstance(value, str):
+            return "->>"
+        else:
+            return "->"
+
+    def visit_operation(self, operation: Operation) -> str:
+        args = [arg.accept(self) for arg in operation.arguments]
+        return f"{operation.operator}({','.join(args)})"
+
+    def visit_comparison(self, comparison: Comparison) -> str:
+        if isinstance(comparison.value, list):
+            return self.visit_operation(
+                Operation(
+                    operator=Operator.AND,
```

Comment:
if comparator is CONTAIN and value is a list, this would turn into "attribute contain first value AND attribute contain second value AND..." — is that intended?

Reply:
Good catch. In an ideal world, `CONTAIN` would behave something like [this](https://www.postgresql.org/docs/current/datatype-json.html#JSON-CONTAINMENT) but practically the LLM seems to use it more for substrings (at least when using OpenAI). Something closer to:
```python
"substring" in my_string
```

Curious what your thoughts are on this. If we are okay treating it as a substring, `AND` is probably the most sane default for situations like:
```python
Comparison(
    attribute='genre',
    comparator=<Comparator.CONTAIN: 'contain'>,
    value=["science", "fiction"],
)
```

Reply:
Simplifying problem by removing `CONTAIN` for now.